### DISCUSSION TITLE
Loading game data from gog

### DISCUSSION THREAD

**Discussion by cxong** [ihhub/fheroes2#4453] [General]
I opened this PR recently: https://github.com/ihhub/fheroes2/pull/4451
It allows fheroes2 to attempt to load homm2 data from a windows gog installation

However fheroes2 already has a script that is run manually that can copy game data, so there is some overlap.

Should the PR be merged? Or changes made before merging? What about the game data copy script?

The advantage of this PR is that users don't have to run the script, the game automatically finds gog data, saving an extra step, and some disk space.
However the script supports many more install locations for homm2, whereas the PR only looks for gog galaxy registry location.

@Branikolog @oleg-derevenetz @ihhub

**Comment by oleg-derevenetz** [comment]
As for me, I personally don't like the idea of integrating such local, non-universal and momentary things into the core engine. If the existing script has any problems with detection of some installations of the original game, it would be better to improve the script.

**Comment by Branikolog** [comment]
Hi, @cxong!
In a case we'd have two different folders, that fheroes engine interacts with, which is not so user friendly. There're already some users being offenced with the fact, engine creates a folder in ".../users" directory.
Also, what will happen if user would delete or move original H2 folder? 
In my opinion is much more convenient to offer user either place fheroes2 into OG folder, or copy needed files from the OG to the fheroes2 folder.

**Reply by cxong** [reply]
> There're already some users being offenced with the fact, engine creates a folder in ".../users" directory.

This PR doesn't create folders

> Also, what will happen if user would delete or move original H2 folder?

The code detects the installation directory from the registry, so if they move the folder via gog settings it will be able to read the correct location. Deleting homm2 will obviously mean fheroes2 can't access it anymore, but I don't see a use case where the user is worse off, other than not wanting to have homm2 installed at the same time as fheroes2 for some reason.

> In my opinion is much more convenient to offer user either place fheroes2 into OG folder, or copy needed files from the OG to the fheroes2 folder.

Yeah these are convenient use cases, and we can keep them, in addition to the PR.

**Reply by Branikolog** [reply]
Is it possible offer three separate ways of installing fheroes2 to user in one dialog during installation?
1) to the OG folder;
2) to a separate folder + copying OG files;
3) to a separate folder + loading files files from OG;
While installing using option "3" we can also note user, that in a case he deletes OG game, fheroes2 won't work anymore.

**Reply by cxong** [reply]
I vote against option 1, as installation of fheroes2 should be separate from homm2 game, and might lead to stray files when users uninstall programs. Also with GOG the user can move the homm2 location, so if they were relying on fheroes2 being in the same folder, after this action it will stop working.

**Reply by ihhub** [reply]
Based on this discussion I have some thoughts:
- since we allow to copy files from the original game directory is it possible to remember the list of files which we copy and then remove them during fheroes2 uninstall? I know that is not an easy way to do so just thoughts in air
- fheroes2 still has an option to add a custom path to be scanned for all data files. Please take a look at `ListDirs Settings::GetRootDirs()` method. We can just specify `FHEROES2_DATA` environment variable with the path to GoG directory which will do the required stuff. We can just declare this variable during installation if GoG directory found.

I also feel that putting OS dependent logic for dealing with original files is not a proper way to do. This job should be done either by scripts or by installation application.

**Reply by cxong** [reply]
Note that setting the variable during install means if the user moves the data folder later, it will not be found. However, this is an uncommon thing I think.

**Comment by lythacore** [comment]
Making it an option during installation would make it less confusing for users - just an opinion from a user.
